### PR TITLE
[Oxli] Initial oxli build 

### PR DESCRIPTION
**Pull request by bocajnotnef**
Initial oxli build using @ctb's API prototype found here: https://github.com/ctb/2015-experimental-graphalign/blob/445c200e662719c2bbf2f172a994ab6806d01edd/khmer_api.py

### CODE REVIEW

**Comment by bocajnotnef** [comment]
Working off of https://image.slidesharecdn.com/2015-vancouver-vanbug-150312150935-conversion-gate01/95/2015-vancouvervanbug-41-638.jpg?cb=1426209146 as a guide(ish)

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
Working in Normalize-by-median, found this:

```
 88         passed_filter = False                                                   
 89         passed_length = True                                                    
 90         for record in batch:                                                    
 91             if len(record.sequence) < ksize:                                    
 92                 passed_length = False                                           
 93                 continue                                                        
 94                                                                                 
 95             seq = record.sequence.replace('N', 'A')                             
 96             med, _, _ = htable.get_median_count(seq)                            
 97                                                                                 
 98             if med < desired_coverage:                                          
 99                 htable.consume(seq)                                             
100                 passed_filter = True                                            
101                                                                                 
102         # Emit records if any passed                                            
103         if passed_length and passed_filter:                                     
104             for record in batch:                                                
105                 write_record(record, outfp)                                     
106         else:                                                                   
107             discarded += batch_size 
```

I'm confused by the comment on 102--seems to indicate if *any* of the records pass the filters then *all* the records in the batch are shoved into the outfp, but reading through the code seems to imply it's actually "if ANY pass the coverage filter and if ALL pass the length filter"

Am I reading this right?

Also, supposing we're doing paired reads, do we output each 'batch' seperately?

@mr-c @ctb

**Comment by mr-c** [comment]
Here batch refers to either a pair or single read (See https://github.com/ged-lab/khmer/blob/2012-paper-diginorm/scripts/normalize-by-median.py#L48)

I agree that reading this block of code without that context is quite confusing which speaks to how messy this function has gotten :-/

File: tests/test_scripts.py [hunk 1]
```diff
@@ -2207,15 +2237,15 @@ def test_sample_reads_randomly():
 
     seqs = set([r.name for r in screed.open(outfile)])
     print seqs
-    assert seqs == set(['850:2:1:1849:7272/1',  '850:2:1:1849:7272/2',
-                        '850:2:1:1574:8683/1',  '850:2:1:1574:8683/2',
+    assert seqs == set(['850:2:1:1849:7272/1', '850:2:1:1849:7272/2',
```
Comment (bocajnotnef) [inline tests/test_scripts.py]: pep8 changes, here

**Comment by bocajnotnef** [comment]
restest this, please

**Comment by bocajnotnef** [comment]
@mr-c CR?

**Comment by bocajnotnef** [comment]
@camillescott @luizirber CR, please?

**Comment by bocajnotnef** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

**Comment by bocajnotnef** [comment]
@camillescott @luizirber CR, please?

File: ChangeLog [hunk 2]
```diff
@@ -18,6 +23,7 @@
    sample-reads-randomly.py.
    * tests/test_scripts.py: updated tests to correspond with correct
    behavior of sample-reads-randomly.py.
+>>>>>>> 97b897f3ce4998da7c97111a7f82ea9f60c3c3b4
```
Comment (luizirber) [inline ChangeLog]: Oops.

File: oxli/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#                                                                               
+# This file is part of khmer, http://github.com/ged-lab/khmer/, and is          
+# Copyright (C) Michigan State University, 2010-2015. It is licensed under
```
Comment (luizirber) [inline oxli/__init__.py]: Just "2015" for the copyright?

File: oxli/khmer_api.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    for n, is_pair, read1, read2 in input_stream:
+        if is_pair:
+            write_record_pair(read1, read2, out_fp)
+        else:
+            write_record(read1, out_fp)
+
+
+def build_graph(ifilenames, graph, num_threads = 1, tags = False):
+    
+    if(tags):
```
Comment (luizirber) [inline oxli/khmer_api.py]: No need for parenthesis.

File: ChangeLog (hunk 2 above)
Comment (bocajnotnef) [inline ChangeLog]: Gaaah. Every time!

File: oxli/khmer_api.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        eat = graph.consume_fasta_and_tag_with_reads_parser
+    else:
+        eat = graph.consume_fasta_with_reads_parser
+    
+    for _, ifile in enumerate(ifilenames):
+        rparser = khmer.ReadParser(ifile)
+        threads = []
+    
+        for num in xrange(num_threads):
+            cur_thread = threading.Thread( target=eat, args=(rparser,))
```
Comment (luizirber) [inline oxli/khmer_api.py]: Remove whitespace before "target"

File: oxli/khmer_api.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            threads.append(cur_thread)
+            cur_thread.start()
+
+        for thread in threads:
+            thread.join()
+        
+
+def diginorm(input_stream, ct, coverage, ksize=0, ofp = False):
```
Comment (luizirber) [inline oxli/khmer_api.py]: Remove whitespaces around "="

File: oxli/khmer_api.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    n = 0
+    discard = 0
+    passed = False
+    for _, is_pair, read1, read2 in input_stream:
+        if is_pair:
+            med1, _, _ = ct.get_median_count(read1.sequence)
+            med2, _, _ = ct.get_median_count(read2.sequence)
+
+            if (med1 < coverage and len(read1.sequence) > ksize) \
+                    or (med2 < coverage and len(read2.sequence) > ksize):
```
Comment (luizirber) [inline oxli/khmer_api.py]: Line continuations can be avoided:
```python
if ((med1 < coverage and len(read1.sequence) > ksize) 
    or (med2 < coverage and len(read2.sequence) > ksize)):
```

File: oxli/khmer_api.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        else:
+            med, _, _ = ct.get_median_count(read1.sequence)
+            if med >= normalize_coverage:
+                read1 = _trim_record(read1, ct, trusted_coverage)
+            yield n, False, read1, None
+            n += 1
+
+
+def streamtrim(input_stream, ct, normalize_coverage, trusted_coverage):
+    import khmer.utils
```
Comment (luizirber) [inline oxli/khmer_api.py]: Why not import `broken_paired_reader` on line https://github.com/ged-lab/khmer/pull/961/files#diff-0fe57386cad3c94ff54bb7daece9b655R12 ?

Comment (bocajnotnef) [inline oxli/khmer_api.py]: to replace the `khmer.utils` import?

Comment (luizirber) [inline oxli/khmer_api.py]: and also https://github.com/ged-lab/khmer/pull/961/files#diff-0fe57386cad3c94ff54bb7daece9b655R174 and https://github.com/ged-lab/khmer/pull/961/files#diff-0fe57386cad3c94ff54bb7daece9b655R175.

Putting all imports in the top of the file is a good practice, also easier to see what are the dependencies.

**Comment by bocajnotnef** [comment]
@luizirber Updated

**Comment by ctb** [comment]
@mr-c and I talked about this today, and came to the following tentative conclusions:

* oxli package itself can/should be created for 2.0;
* you should create the 'oxli' entry point command 'oxli build-graph' to run this code;
* you can also replace 'load-graph.py' with this code, but please don't create a new entry point;
* the unused bits khmer_api.py should be removed;

@bocajnotnef any thoughts?

**Comment by bocajnotnef** [comment]
For the oxli entry point,  I assume you mean being able to run build-graph from the command line?

Not sure what other entry point I would be making (i.e., not sure what I'm avoiding)

**Comment by ctb** [comment]
On Mon, May 18, 2015 at 01:06:12PM -0700, Jake Fenton wrote:
> For the oxli entry point,  I assume you mean being able to run build-graph from the command line?
> 
> Not sure what other entry point I would be making (i.e., not sure what I'm avoiding)

% oxli build-graph

should be how you call build-graph - the subcommand format.

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
Note: code coverage tests don't include the `oxli/` directory.

**Comment by mr-c** [comment]
You should fix that :-)

**Comment by bocajnotnef** [comment]
@camillescott @luizirber @mr-c CR, please?

File: khmer/_version.py [hunk 9]
```diff
@@ -114,7 +114,7 @@ def git_versions_from_keywords(keywords, tag_prefix, verbose=False):
         # "stabilization", as well as "HEAD" and "master".
         tags = set([r for r in refs if re.search(r'\d', r)])
         if verbose:
-            print("discarding '%s', no digits" % ",".join(refs-tags))
+            print("discarding '%s', no digits" % ",".join(refs - tags))
```
Comment (ctb) [inline khmer/_version.py]: I think these modifications can be discarded. Frankly, you should stop running autopep :)

File: oxli/__init__.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    parser = argparse.ArgumentParser(
+        description='Single entry point script for khmer',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    subparsers = parser.add_subparsers()
+
+    # build-graph (formerly load-graph) parsers here
+    parser_build_graph = \
+        subparsers.add_parser('build_graph',
```
Comment (ctb) [inline oxli/__init__.py]: let's make this build-graph (- instead of _)

File: oxli/common.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
+#
```
Comment (ctb) [inline oxli/common.py]: Why not just import these from khmer for now, instead of duplicating the code?

File: scripts/build-graph.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+#! /usr/bin/env python2
```
Comment (ctb) [inline scripts/build-graph.py]: Please rename this file to load-graph.py

File: scripts/build-graph.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import screed
+from khmer.khmer_args import build_hashbits_args
+from khmer.khmer_args import (report_on_config, info, add_threading_args)
+from khmer.kfile import check_input_files, check_space
+from khmer.kfile import check_space_for_hashtable
+from oxli import khmer_api, build_graph
+
+
+def get_parser():
+    parser = build_hashbits_args(descr="Load-Graph called.")
```
Comment (ctb) [inline scripts/build-graph.py]: descr?

File: scripts/normalize-by-median.py [hunk 14]
```diff
@@ -30,6 +30,7 @@
 from khmer.kfile import (check_space, check_space_for_hashtable,
                          check_valid_file_exists)
 from khmer.utils import write_record, check_is_pair
+
```
Comment (ctb) [inline scripts/normalize-by-median.py]: please don't introduce unneeded /unnecessary changes.

File: tests/test_scripts.py [hunk 15]
```diff
@@ -1016,6 +1016,66 @@ def test_load_graph():
     assert x == (1, 0), x
 
 
+def test_build_graph():
```
Comment (ctb) [inline tests/test_scripts.py]: most of these tests are copies of the tests that are being used on load-graph, yes? in which case they can be removed.

**Comment by ctb** [comment]
Done. good stuff, but still things to be cleaned up! let me know if you want rationales/reasons/explanations/more info on anything.

File: oxli/common.py (hunk 11 above)
Comment (bocajnotnef) [inline oxli/common.py]: I had to change a few of the `build_parser` funcs to add to a parser instead of generate one (because of how subparsers work).

I could modify the khmer ones to be along the line of
```
build_hashbits(parser=None):
   if parser == None:

```

File: tests/test_scripts.py (hunk 15 above)
Comment (bocajnotnef) [inline tests/test_scripts.py]: Indeed, especially if I'm removing `scripts/build-graph.py`. Or moving it back into `load-graph`, which amounts to the same thing.

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
@mr-c I'm getting a strange test failure here--the linux host is dying on the sphinx docgen step but I can't reproduce the problem on Athyra.

**Comment by mr-c** [comment]
You need the unreleased version of sphinx-contrib/autoprogram. Look at the
Jenkins build script and doc/requirements.txt

**Comment by bocajnotnef** [comment]
I'm having trouble determining what's causing the issue--Everything was working fine until the last couple of commits but I don't see what I changed that could cause this issue. From looking at the error it looks like sphinx is dying on the `setup()` line in setup.py but I didn't edit that in these commits.

**Comment by bocajnotnef** [comment]
Figured out it's because the versioneer cmdclass returns a list when a buffer is expected, but still not getting why this is just now a problem.

Also not sure which cmd class to give it.

**Comment by bocajnotnef** [comment]
Not the versioneer cmdclass--No idea what it is now. Looks to be failing on `**SETUP_METADATA` but the only changes made to that were done by Luiz and were merged in by 3bb86af--which passed.

File: scripts/load-graph.py [hunk 16]
```diff
@@ -18,105 +18,22 @@
...
-                        ' k-mer presence table filename.')
-    parser.add_argument('input_filenames', metavar='input_sequence_filename',
-                        nargs='+', help='input FAST[AQ] sequence filename')
-    parser.add_argument('--report-total-kmers', '-t', action='store_true',
-                        help="Prints the total number of k-mers to stderr")
-    parser.add_argument('--write-fp-rate', '-w', action='store_true',
-                        help="Write false positive rate into .info file")
-    parser.add_argument('-f', '--force', default=False, action='store_true',
-                        help='Overwrite output file if it exists')
+    parser = build_hashbits_args()
```
Comment (mr-c) [inline scripts/load-graph.py]: Why are you no longer passing in the `descr`?

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
@mr-c @ctb @camillescott @luizirber Updated; CR, please?

File: khmer/khmer_args.py [hunk 17]
```diff
@@ -60,9 +60,12 @@ def build_counting_args(descr=None, epilog=None):
...
 
 
-def build_hashbits_args(descr=None, epilog=None):
+def build_hashbits_args(descr=None, epilog=None, parser=None):
     """Build an ArgumentParser with args for hashbits based scripts."""
-    parser = build_hash_args(descr=descr, epilog=epilog)
+    if parser is None:
+        parser = build_hash_args(descr=descr, epilog=epilog)
+    else:
+        parser = build_hash_args(descr=descr, epilog=epilog, parser=parser)
```
Comment (luizirber) [inline khmer/khmer_args.py]: since `build_hash_args` has the same default for `parser`, you can just pass it and avoid checking if it is None.

File: oxli/build_graph.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+# This file is part of khmer, http://github.com/ged-lab/khmer/, and is
+# Copyright (C) Michigan State University, 2009-2015. It is licensed under
+# the three-clause BSD license; see doc/LICENSE.txt.
+# Contact: khmer-project@idyll.org
+#
+# pylint: disable=invalid-name,missing-docstring
+"""
+Build a graph from the given sequences, save in <ptname>.
+
+% python scripts/load-graph.py <ptname> <data1> [ <data2> <...> ]
```
Comment (luizirber) [inline oxli/build_graph.py]: This isn't true anymore

File: oxli/__init__.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    khmer_args.build_hashbits_args("Load sequences into the compressible"
+                                   "graph format plus optional tagset.",
+                                   None, parser=parser_build_graph)
+    build_graph.build_parser(parser_build_graph)
+    parser_build_graph.set_defaults(func=execute_build_graph)
+
+    return parser
+
+
+def execute_build_graph(args):
```
Comment (mr-c) [inline oxli/__init__.py]: Why does this function exist?

File: oxli/__init__.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    build_graph.main(args)
+
+
+def main():
+    """
+    main function; does the parsing and kicks off the subcommand
+    """
+    args = get_parser().parse_args()
+    args.func(args)
+    return 0
```
Comment (ctb) [inline oxli/__init__.py]: is the return value ever used?

File: oxli/build_graph.py (hunk 18 above)
Comment (bocajnotnef) [inline oxli/build_graph.py]: No, that still works--@ctb had me remove `build-graph.py` so that will work.
Granted, the oxli command also works but I don't think we're officially moving to that yet.

File: oxli/khmer_api.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
+#
```
Comment (ctb) [inline oxli/khmer_api.py]: Suggest renaming this from 'khmer_api' to something else. Not sure what - maybe just 'api' or 'functions'?

File: scripts/load-graph.py [hunk 22]
```diff
@@ -2,7 +2,7 @@
 #
 # This file is part of khmer, https://github.com/dib-lab/khmer/, and is
 # Copyright (C) Michigan State University, 2009-2015. It is licensed under
-# the three-clause BSD license; see LICENSE.
+# the three-clause BSD license; see doc/LICENSE.txt.
```
Comment (ctb) [inline scripts/load-graph.py]: FIXME

File: oxli/khmer_api.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+# Contact: khmer-project@idyll.org
+#
+# pylint: disable=missing-docstring
+
+
+import threading
+import khmer.utils
+
+
+def build_graph(ifilenames, graph, num_threads=1, tags=False):
```
Comment (mr-c) [inline oxli/khmer_api.py]: What is the point of this?

**Comment by ctb** [comment]
LGTM in concept!

File: oxli/khmer_api.py (hunk 23 above)
Comment (mr-c) [inline oxli/khmer_api.py]: Nevermind, I found the call in `oxli/build_graph.py`

File: oxli/__init__.py (hunk 20 above)
Comment (bocajnotnef) [inline oxli/__init__.py]: Actually yes. There's a strange issue @luizirber helped me troubleshoot where if we don't explicitly return something the testing framework grabs `None` as the status code and dies.

File: oxli/khmer_api.py (hunk 23 above)
Comment (mr-c) [inline oxli/khmer_api.py]: naming inputs with a 'i' prefix is not a convention we use. Also needs docstrings.

Comment (bocajnotnef) [inline oxli/khmer_api.py]: The point of what? That is build_graph, now.

File: oxli/__init__.py (hunk 20 above)
Comment (luizirber) [inline oxli/__init__.py]: I suggested this, because the way we run scripts for tests. Full discussion on the `ged-lab/dev-flood` gitter channel, but:

What I think it’s happening here is that we don’t have a ‘pure’ shell running the script, so when we run from the shell the None returned is converted to 0
but in our runscript() it’s actually inside Python
and the return code just stays None

Comment (luizirber) [inline oxli/__init__.py]: Super old post by Guido: https://www.artima.com/weblogs/viewpost.jsp?thread=4829
Recent one, good tips: http://www.jeffknupp.com/blog/2014/02/04/starting-a-python-project-the-right-way/

We can remove the `return 0`, but we might want to replace the `main()` call for `sys.exit(main())`

Comment (ctb) [inline oxli/__init__.py]: > We can remove the `return 0`, but we might want to replace the `main()` call for `sys.exit(main())`

+1

**Comment by bocajnotnef** [comment]
@luizirber @ctb Cleaned up--Ready for merge?

File: oxli/build_graph.py (hunk 18 above)
Comment (ctb) [inline oxli/build_graph.py]: correct.

File: oxli/build_graph.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+                        action='store_true', dest='no_build_tagset',
+                        help='Do NOT construct tagset while loading sequences')
+    parser.add_argument('output_filename',
+                        metavar='output_presence_table_filename', help='output'
+                        ' k-mer presence table filename.')
+    parser.add_argument('input_filenames', metavar='input_sequence_filename',
+                        nargs='+', help='input FAST[AQ] sequence filename')
+    parser.add_argument('--report-total-kmers', '-t', action='store_true',
+                        help="Prints the total number of k-mers to stderr")
+    parser.add_argument('--write-fp-rate', '-w', action='store_true',
```
Comment (ctb) [inline oxli/build_graph.py]: turn on by default ref #1011

File: oxli/build_graph.py [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+    add_threading_args(parser)
+    parser.add_argument('--no-build-tagset', '-n', default=False,
+                        action='store_true', dest='no_build_tagset',
+                        help='Do NOT construct tagset while loading sequences')
+    parser.add_argument('output_filename',
+                        metavar='output_presence_table_filename', help='output'
+                        ' k-mer presence table filename.')
+    parser.add_argument('input_filenames', metavar='input_sequence_filename',
+                        nargs='+', help='input FAST[AQ] sequence filename')
+    parser.add_argument('--report-total-kmers', '-t', action='store_true',
```
Comment (ctb) [inline oxli/build_graph.py]: turn on by default ref #1011

Comment (ctb) [inline oxli/build_graph.py]: (by which I mean "remove as option" because I don't imagine anyone will not want to know this)

File: oxli/build_graph.py [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def main(args):
+    info('build-graph.py', ['graph', 'SeqAn'])
+
+    report_on_config(args, hashtype='hashbits')
+    base = args.output_filename
+    filenames = args.input_filenames
+
+    for _ in args.input_filenames:
```
Comment (ctb) [inline oxli/build_graph.py]: replace '_' with a named variable.

File: scripts/load-graph.py [hunk 27]
```diff
@@ -22,101 +22,19 @@
...
-    if not args.no_build_tagset:
-        print >> sys.stderr, 'and ' + base + '.tagset'
+    parser = build_graph.build_parser(parser)
+    return parser
 
 if __name__ == '__main__':
-    main()
+    parser = get_parser()
+    args = parser.parse_args()
+    build_graph.main(args)
```
Comment (ctb) [inline scripts/load-graph.py]: sys.exit(...)

**Comment by ctb** [comment]
Before merge, this is a good candidate for squashing, ref #1013.

File: oxli/build_graph.py (hunk 24 above)
Comment (bocajnotnef) [inline oxli/build_graph.py]: turn on by default as in always on with no way to disable, or flip flop the arguments so the arg disables reporting?

Can't imagine why you'd want to turn that off, though...

Comment (mr-c) [inline oxli/build_graph.py]: No way to disable

File: oxli/__init__.py (hunk 20 above)
Comment (bocajnotnef) [inline oxli/__init__.py]: So, am I removing the return? If so, what am I adding in?

Comment (luizirber) [inline oxli/__init__.py]: `>>> sys.exit(None)`
will have a return code 0 for the shell, so you can just remove it.

File: oxli/__init__.py [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+
+if __name__ == '__main__':
+    main()
```
Comment (luizirber) [inline oxli/__init__.py]: here you do
```sys.exit(main())```

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
@ctb @luizirber Updated; CR?

**Comment by ctb** [comment]
LGTM; thanks!

**Comment by ctb** [comment]
@bocajnotnef or others - please merge if the build succeeds (which it should - I just fixed the ChangeLog and updated to latest master).
